### PR TITLE
Fway/feat/tdp 10867/filter valid invalid

### PR DESCRIPTION
**Pull request by fway87**
The purpose of this Pull Request is to add three custom functions for DSEL : 'isOfType', 'isValid' and 'isInvalid'.
For this purpose, a new Daikon module is built and include this custom functions.
 
**Link to the JIRA issue**
https://jira.talendforge.org/browse/TDP-10867
 
**Please check if the Pull Request fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/Talend/daikon/blob/master/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features, coverage should be over 75% in the new code)
- [x] Docs have been added / updated (for bug fixes / features)
- [ ] [NOT NEEDED HERE] Related design / discussions / pages (not in JIRA), if any, are all linked or available in the Pull Request

### CODE REVIEW

File: daikon-dsel/src/test/java/org/talend/dsel/functions/custom/IsOfTypeTest.java [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+        CategoryRegistryManager categoryRegistryManager = new CategoryRegistryManager(properties);
+        DeletableDictionarySnapshotOpener opener = new DeletableDictionarySnapshotOpener(properties,
+                categoryRegistryManager.getSharedDictionary());
+        DQCategory integerCategory = createIntegerCategory();
+        CustomDictionaryHolder holder = categoryRegistryManager.getCustomDictionaryHolder("fakeTenantId");
+
+        holder.createCategory(integerCategory);
+        dictionarySnapshot = opener.openDeletableDictionarySnapshot("fakeTenantId");
+
+        HPathStore store = new HPathMapStore();
```
Comment (abenali-talend) [inline daikon-dsel/src/test/java/org/talend/dsel/functions/custom/IsOfTypeTest.java:349]: same

File: daikon-dsel/src/main/java/org/talend/dsel/exception/FunctionException.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package org.talend.dsel.exception;
+
+public class FunctionException extends RuntimeException {
+
+    public FunctionException(String msg) {
+        super(msg);
+    }
+
+    public FunctionException(String msg, Throwable cause) {
```
Comment (abenali-talend) [inline daikon-dsel/src/main/java/org/talend/dsel/exception/FunctionException.java:9]: This constructor seems not be used

File: daikon-dsel/src/test/java/org/talend/dsel/functions/custom/FunctionTest.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            interpreter.eval();
+        });
+    }
+
+    protected void testEvalExpression(boolean expectedResult, String expression) {
+        interpreter.setExpression(expression);
+        assertThat(interpreter.evalAsBoolean()).isEqualTo(expectedResult);
+    }
+
+    protected void testEvalExpression(boolean expectedResult, String expression, String varName, Object varValue) {
```
Comment (abenali-talend) [inline daikon-dsel/src/test/java/org/talend/dsel/functions/custom/FunctionTest.java:31]: This method will be used in the future ?

File: daikon-dsel/src/test/java/org/talend/dsel/functions/custom/FunctionTest.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+public abstract class FunctionTest {
+
+    protected ExprInterpreter interpreter;
+
+    @BeforeEach
+    public void init() {
+        interpreter = ExprInterpreterFactory.create();
+    }
+
+    protected void testError(Class<? extends Exception> expectedException, String expression) {
```
Comment (abenali-talend) [inline daikon-dsel/src/test/java/org/talend/dsel/functions/custom/FunctionTest.java:19]: This method will be used in the future ?

File: daikon-dsel/src/test/java/org/talend/dsel/functions/custom/IsValidTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    public static void setUp() {
+        SemanticProperties properties = new SemanticProperties(tempDir.toString());
+        CategoryRegistryManager categoryRegistryManager = new CategoryRegistryManager(properties);
+        DeletableDictionarySnapshotOpener opener = new DeletableDictionarySnapshotOpener(properties,
+                categoryRegistryManager.getSharedDictionary());
+        dictionarySnapshot = opener.openDeletableDictionarySnapshot("fakeTenantId");
+
+        isValid = spy(IsValid.class);
+
+        HPathStore store = new HPathMapStore();
```
Comment (abenali-talend) [inline daikon-dsel/src/test/java/org/talend/dsel/functions/custom/IsValidTest.java:44]: HPathMapStore is deprecated, may be you should use DselHPathStore

File: daikon-dsel/src/main/java/org/talend/dsel/exception/FunctionException.java (hunk 2 above)
Comment (fway87) [reply daikon-dsel/src/main/java/org/talend/dsel/exception/FunctionException.java:9]: Yes, never used for now, so is useless. I delete it ;)

File: daikon-dsel/src/test/java/org/talend/dsel/functions/custom/IsOfTypeTest.java [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+    public static void setUp() {
+        SemanticProperties properties = new SemanticProperties(tempDir.toString());
+        CategoryRegistryManager categoryRegistryManager = new CategoryRegistryManager(properties);
+        DeletableDictionarySnapshotOpener opener = new DeletableDictionarySnapshotOpener(properties,
+                categoryRegistryManager.getSharedDictionary());
+        dictionarySnapshot = opener.openDeletableDictionarySnapshot("fakeTenantId");
+
+        isOfType = spy(IsOfType.class);
+
+        HPathStore store = new HPathMapStore();
```
[the remark this answers is not in the document]
Comment (fway87) [reply daikon-dsel/src/test/java/org/talend/dsel/functions/custom/IsOfTypeTest.java:57]: Yep, it's better to not wait for it to disappear soon :+1:

File: daikon-dsel/src/test/java/org/talend/dsel/functions/custom/IsOfTypeTest.java (hunk 1 above)
Comment (fway87) [reply daikon-dsel/src/test/java/org/talend/dsel/functions/custom/IsOfTypeTest.java:349]: yes

File: daikon-dsel/src/test/java/org/talend/dsel/functions/custom/FunctionTest.java (hunk 4 above)
Comment (fway87) [reply daikon-dsel/src/test/java/org/talend/dsel/functions/custom/FunctionTest.java:19]: I will add some tests for the error cases for each of the custom functions in this PR, so yes this method will be used finally :)

File: daikon-dsel/src/test/java/org/talend/dsel/functions/custom/FunctionTest.java (hunk 3 above)
Comment (fway87) [reply daikon-dsel/src/test/java/org/talend/dsel/functions/custom/FunctionTest.java:31]: Nope, this method is useless, I delete it in this PR

File: daikon-dsel/src/test/java/org/talend/dsel/functions/custom/IsValidTest.java (hunk 5 above)
Comment (fway87) [reply daikon-dsel/src/test/java/org/talend/dsel/functions/custom/IsValidTest.java:44]: Yep

**Comment by fway87** [comment]
I did all the changes, and also did a cleanup in the pom.xml (removing the useless testcontainers dependancies and some useless properties).

**Review by abenali-talend — APPROVED** [review]

**Pull request merged**